Allows field generators to contain explosions.

Pickle-Coding opened the pull request:
## About The Pull Request
Allows field generators to contain explosions. The fields will not let the explosion propagate past them.
## Why It's Good For The Game
Allows people to contain the tesla for fun, or on station bomb testing if they so desire.
## Changelog
:cl:
balance: Field generators can parry explosions.
/:cl:

Pickle-Coding committed:
Field generator bomb resistance.

Gives 100% bomb armour to field generator armor datum.

Pickle-Coding committed:
Field generator block explosives.

Gives field generators the block explosive element, allowing them to have explosive blocking behaviour. block_resist gets set to INFINITY when set up, 0 when cleaned up.

Pickle-Coding committed:
Containment field explosion block.

Gives the containment fields an explosion_block of INFINITY, and the blocks_explosives element.

Pickle-Coding committed:
Erm, I hope this works??

Windows did it like this.

LemonInTheDark commented:
Have you tested this locally?

LemonInTheDark commented on code/modules/power/singularity/field_generator.dm:
```diff
@@ -72,9 +72,9 @@ no power level overlay is currently in the overlays list.
+	AddElement(/datum/element/blocks_explosives)
 	. = ..()
 	AddElement(/datum/element/empprotection, EMP_PROTECT_SELF | EMP_PROTECT_WIRES)
-	AddElement(/datum/element/blocks_explosives)
```
I can't really test post init, which I would prefer to here. I don't really like that this is forced, even tho it really should not be

Pickle-Coding replied:
I'm not really sure why it's like this. Windows adds the element before parent call so I'm just doing that to hopefully pass the tests.

LemonInTheDark replied:
The element is required, and I need to test in init to ensure we never miss a block add.
I don't have a good way to like.... check after each atom's init tho, so I can't do what I want and let you just do what you wanted to. I'm sorry

Pickle-Coding replied:
Not really hindering me, just looks kinda weird.

Pickle-Coding commented:
It's 6 AM. I'll test this later.

Pickle-Coding committed:
Add element before parent call.

Add the blocks_explosives element before parent call. Hoepfully this makes it work properly.

Jackal-boop commented:
I proposed this idea a year ago I should be given royalties.

Pickle-Coding commented:
Do you want to be given some of the negative GBP?

Pickle-Coding commented:
Tested, and it works.

![tesla png](https://github.com/tgstation/tgstation/assets/58013024/385e9c02-a0ac-4ced-b8a4-4449ca2fa1b5)

moocowswag commented on code/modules/power/singularity/containment_field.dm:
```diff
@@ -14,12 +14,14 @@
 
 /obj/machinery/field/containment/Initialize(mapload)
+	AddElement(/datum/element/blocks_explosives)
 	. = ..()
```
Why does the addelement go before this?

Pickle-Coding replied:
It's to pass unit tests, which I think occurs somewhere in the parent initialize calls. Can't detect the element if it was added after the parent class initialize calls.

moocowswag replied:
lol i looked at the code for obj/structure/window and it also does this

LemonInTheDark replied:
I am so sorry

moocowswag commented:
Test if the explosion block remains on the turfs after the containment field goes down real quick because im testing similar code and its not going away when my shields go down.

Jackal-boop commented:
> Do you want to share some of the negative GBP reward with me?

Yes
